django/db/models/sql/query.py

```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
         key field for example).
         """
         joins = [alias]
-        # First, generate the path for the names
-        path, final_field, targets, rest = self.names_to_path(
-            names, opts, allow_many, fail_on_missing=True)
+        transforms = []
+        # Try resolving all the names as fields first, but if there is an error
+        # progressively treat trailing names as lookups until we can resolve
+        # a field
+        last_field_exception = None
+        for pivot in range(len(names), 0, -1):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
+            except (FieldError) as exc:
```

Comment:
Thank you, but as I said in the description I'm lokking for help with the mewt of the problem right now, not style advice. This does currently contain WIP code that is not suitable for review.